### PR TITLE
Develop search book controller

### PR DESCRIPTION
**Pull request by alwayswannajava**

### CODE REVIEW

File: src/main/java/com/spring/onlinebookstore/repository/book/specification/TitleSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey(String key) {
+        return "title";
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("title").in(Arrays.stream(params).toArray());
```
Comment (okuzan) [inline src/main/java/com/spring/onlinebookstore/repository/book/specification/TitleSpecificationProvider.java]: use constants

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/spring/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -32,4 +35,15 @@ public BookDto getBookById(@PathVariable int id) {
...
         return bookService.save(bookRequestDto);
     }
+
+    @DeleteMapping("/{id}")
+    public void deleteBook(@PathVariable int id) {
+        bookService.deleteById(id);
+    }
+
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable long id,
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/controller/BookController.java]:
```suggestion
    public BookDto updateBook(@PathVariable Long id,
```

File: src/main/java/com/spring/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(SearchBookRequestDto searchBookRequestDto) {
+        Specification<Book> specification = Specification.where(null);
+        if (searchBookRequestDto.author() != null && searchBookRequestDto.author().length > 0) {
+            specification = specification.and(specificationProviderManager
+                    .getSpecificationProvider("author")
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/repository/book/BookSpecificationBuilder.java]: create constants for author, title, isbn

File: src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java [hunk 4]
```diff
@@ -30,9 +31,25 @@ public List<BookDto> findAll() {
     }
 
     @Override
-    public BookDto findById(int id) {
+    public BookDto findById(long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java]:
```suggestion
    public BookDto findById(Long id) {
```

File: src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java [hunk 5]
```diff
@@ -30,9 +31,25 @@ public List<BookDto> findAll() {
...
         Book book = bookRepository.findById(id).orElseThrow(() ->
                 new EntityNotFoundException("Can't find book by id: " + id));
         return bookMapper.toDto(book);
     }
+
+    @Override
+    public void deleteById(long id) {
+        Book book = bookRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("Can't find book by id: " + id));
+        bookRepository.deleteById(book.getId());
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/service/BookServiceImpl.java]:
```suggestion
    public void deleteById(Long id) {
        bookRepository.deleteById(id);
```

File: src/main/java/com/spring/onlinebookstore/service/BookService.java [hunk 6]
```diff
@@ -2,12 +2,17 @@
...
     BookDto save(CreateBookRequestDto requestDto);
 
     List<BookDto> findAll();
 
-    BookDto findById(int id);
+    BookDto findById(long id);
+
+    void deleteById(long id);
+
+    BookDto update(long id, UpdateBookRequestDto updateBookRequestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/spring/onlinebookstore/service/BookService.java]:
```suggestion
    BookDto findById(Long id);

    void deleteById(Long id);

    BookDto update(Long id, UpdateBookRequestDto updateBookRequestDto);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Pull request closed without merging**